### DISCUSSION TITLE
Automatically determine chain by chainId.

### DISCUSSION THREAD

**Discussion by benesjan** [wevm/viem#418] [Question] [answered]
I need to determine the correct chain only based on the RPC URL. Is there some way to make viem ask for the chainId and then set the correct chain automatically (e.g. if chain id = 1 set mainnet ...) or do I have to write this functionality on my own?

And if I need to do this on my own can use a publicClient for this?

**Comment by jxom** [comment, marked as answer]
Yep, you will need to do this on your own, and create your own mapping before passing the chain to viem:

```tsx
import * as chains from 'viem/chains'

function getChain(id: number) {
  return chains.find(x => x.id === id)
}

const id = getId() 
const chain = getChain(id)
```

**Reply by fubhy** [reply]
Note that the wildcard import will literally pull in *every single* chain into your final app bundle. That might not matter as much if you are building a backend service but If it's a frontend app, you might want to consider putting the subset of supported chains of your app in a new mapping.

**Comment by benesjan** [comment]
Thank you all for the responses. Here is the final code I used:

```
import * as chains from 'viem/chains';

...

  /**
   * Gets the chain object for the given chain id.
   * @param chainId - Chain id of the target EVM chain.
   * @returns Viem's chain object.
   */
  private getChain(chainId: number) {
    for (const chain of Object.values(chains)) {
      if ('id' in chain) {
        if (chain.id === chainId) {
          return chain;
        }
      }
    }

    throw new Error(`Chain with id ${chainId} not found`);
  }
```

I couldn't use `chains.find(x => x.id === id)` because `chains` is an object and not an array that can be iterated over. I am using viem on a backend service so pulling all the chains in the final app bundle is ok.

**Comment by fubhy** [comment]
@jxom To make use cases like this easier, I'd suggest to remove `defineChain` from the exports of `viem/chain` so that the wildcard import is properly typed as a map of `Chain`.

@benesjan To get around that (so you can optimize this fn a bit), you could do this for now:

```ts
import * as all from "viem/chains";

const { defineChain: _, ...chains } = all;

/**
 * Gets the chain object for the given chain id.
 * @param chainId - Chain id of the target EVM chain.
 * @returns Viem's chain object.
 */
function getChain(chainId: number) {
  for (const chain of Object.values(chains)) {
    if (chain.id === chainId) {
      return chain;
    }
  }

  throw new Error(`Chain with id ${chainId} not found`);
}
```

**Reply by fubhy** [reply]
This should be better imho: #424

**Reply by 0x33dm** [reply]
@fubhy what do you mean "this should be better" ? I don't see a different implementation in the link you have posted?

**Reply by jxom** [reply]
The difference is that `defineChain` is no longer exported from `viem/chains`.

**Comment by bguiz** [comment]
Use [`extractChain`](https://viem.sh/docs/utilities/extractChain.html), seems to accomplish the same thing as this answer:
https://github.com/wevm/viem/discussions/418#discussioncomment-5740428

**Reply by emmaodia** [reply]
Yes, you are right @bguiz
